### ISSUE TITLE
Adversarial robots?

### ISSUE THREAD

**Issue by HuwCampbell** [byorgey/swarm#89]
This may be premature, but is there any thought to add external robots or actors with bad intentions? (or even native wildlife)

**Describe the solution you'd like**
I would like to see survival mode have an adversarial component with things which want to kill you.

**Describe alternatives you've considered**
The world is unhabited and desolate of intelligent life.

**Additional context**
I've seen a few people play minecraft, kids seem to enjoy creative mode, but adults like a bit more of a challenge in survival. I think to make this stick, there may need to be a "danger" mode.

**Comment by byorgey** [comment]
Yes! I totally agree, there should at least be the option to play in a mode like this.  Once we do #29 we will have a framework for generating non-player-controlled robots that spawn as part of the world.  (When I say "robots" I just mean it in the technical sense of entities that run code, not that they will be described or presented to the player as such.)  I have a few ideas for some (non-malicious) native wildlife, but we can definitely brainstorm ideas for adversarial things.

**Comment by TristanCacqueray** [comment]
Similar to The Martian movie, the weather could also become a challenge, for example by affecting the solar panels. Though it might be too much to implement such life support management, that's just an idea for the brainstorm.

**Comment by byorgey** [comment]
Related: #218 .

**Comment by byorgey** [comment]
One minor thing to note is that in order to support this (and in particular, in order to support this as an *optional* feature of the world, so a player can choose to play with or without hostile entities), we would need to support the ability to have different world functions, and have those different functions somehow referenced in a scenario (see #296 ).  I don't think we should have the ability to write arbitrary lambdas etc. in a scenario description though; it's probably fine if there is just a fixed enumeration of possible world functions to choose from, or to take it one step up, perhaps there can be a fixed enumeration of modular world function components to choose from, and one could specify several and have them combined in some way (this idea is also referenced here: https://github.com/swarm-game/swarm/issues/29#issuecomment-1101920998 ).

**Comment by byorgey** [comment]
Some ideas/questions:

- A `cat` which is mostly harmless but wanders around randomly, occasionally "playing with" (*i.e.* moving) items.  This would be mostly annoying but could mess things up for you if you have a carefully crafted environment that your robots depend on.
- In https://github.com/swarm-game/swarm/pull/416#issuecomment-1159323413 I had the idea of a `balrog` that wanders around setting everything on fire.
- Should adversarial things be able to literally destroy robots?  Actually destroying robots outright is pretty harsh.  We could potentially think about giving robots a certain amount of "health" so that there could be a spectrum of attacks ranging from mild to severe, that don't necessarily destroy your robots immediately.

**Comment by xsebek** [comment]
A truly evil robot would be a doppelgänger:
- destroys your robot
- changes it’s looks to the robots
- simulates the robots program until it finds another robot

It could appear once your robot disturbs another crashed spaceship.

**Comment by xsebek** [comment]
I would not like casually passing by aliens to destroy my robots, but if they only spawned later in game, then one shot kills are fine by me.

In Factorio health has more uses - you can hit rocks with car, aliens can damage you, robots can deteriorate - but its visible in the GUI and there are upgrades to increase it. You also get an alert and can run there to save the day.

I am not sure Swarm is well suited to health which to me goes better in GUI action games.

Crippling the robots by breaking  stuff in inventory/installed devices would be easier to fit in current automation of robots where inventory is the centre IMO.

**Comment by byorgey** [comment]
See also #1539 .
